contracts/pallet-ibc/src/ics20/mod.rs

```diff
@@ -561,11 +561,187 @@ pub fn full_ibc_denom(packet: &Packet, mut token: PrefixedCoin) -> String {
+			Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 0);
+			Ics20Error::implementation_specific(format!(
+				"Failed to parse memo : {:?} ",
+				packet_data.memo
+			))
+		})?;
+
+		let prefixed_coin = if is_receiver_chain_source(
+			packet.source_port.clone(),
+			packet.source_channel,
+			&packet_data.token.denom,
+		) {
+			let prefix = TracePrefix::new(packet.source_port.clone(), packet.source_channel);
+			let mut c = packet_data.token.clone();
+			c.denom.remove_trace_prefix(&prefix);
+			c
+		} else {
+			let prefix =
+				TracePrefix::new(packet.destination_port.clone(), packet.destination_channel);
+			let mut c = packet_data.token.clone();
+			c.denom.add_trace_prefix(prefix);
+			c
+		};
+
+		// At this point the asset SHOULD exist
+		let asset_id =
+			<T as crate::Config>::IbcDenomToAssetIdConversion::from_denom_to_asset_id(
+				&prefixed_coin.denom.to_string(),
+			)
+			.map_err(|_| {
+				log::warn!(target: "pallet_ibc", "Asset does not exist for denom: {}", prefixed_coin.denom.to_string());
+				Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 1);
+				Ics20Error::implementation_specific("asset does not exist".to_string())
+			})?;
+
+		let amount = packet_data.token.amount.as_u256().low_u128();
+
+		let raw_bytes = memo.receiver.as_bytes().to_vec();
+
+		let account_id =
+			crate::MultiAddress::<<T as frame_system::Config>::AccountId>::Raw(raw_bytes.clone());
+		let origin = RawOrigin::Signed(receiver.clone());
+		let channel_id = memo.channel.parse().map_err(|_| {
+			Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 2);
+			Ics20Error::implementation_specific("Failed to parse channel ID".to_string())
+		})?;
+
+		//timestamp that current timestamp + 30 minutes
+		let timestamp = <pallet_timestamp::Pallet<T> as frame_support::traits::UnixTime>::now()
+			.as_secs() + 1800;
+
+		let params = crate::TransferParams::<<T as frame_system::Config>::AccountId> {
+			to: account_id,
+			source_channel: channel_id, /* TODO source? should it be destination channel id from
+			                             * memo? */
+			timeout: ibc_primitives::Timeout::Offset {
+				timestamp: Some(timestamp), /* TODO ? what timestamp to use. now is current + 30
+				                             * minutes */
+				height: None, //TODO ? what height
+			},
+		};
+
+		let memo = match memo.next {
+			Some(e) => Some(*e),
+			None => None,
+		};
+
+		let mut next_memo: Option<T::MemoMessage> = None;
+		if let Some(memo) = memo {
+			let memo_str = serde_json::to_string(&memo).map_err(|_| {
+				Self::emit_memo_execution_failed_event(
+					receiver.clone(),
+					packet_data.memo.clone(),
+					3,
+				);
+				Ics20Error::implementation_specific("failed to serialize memo".to_string())
+			})?;
+			let memo_result =
+				<T as crate::Config>::MemoMessage::from_str(&memo_str).map_err(|_| {
```

Comment:
I guess, it's safe to use the original memo, because it was deserialised successfully

Reply:
We can't use original memo. we need to remove parent(current chain) because we execute multihop logic reading memo and should put modified memo for next chains

Reply:
Yes, but the memo is already extracted [here](https://github.com/ComposableFi/centauri/blob/68e96e4aa35f997ac4243d25f067b7aac1208fd0/contracts/pallet-ibc/src/ics20/mod.rs#L786), so I suppose we can use it as is, without [converting](https://github.com/ComposableFi/centauri/blob/68e96e4aa35f997ac4243d25f067b7aac1208fd0/contracts/pallet-ibc/src/ics20/mod.rs#L790-L808) to JSON string and back.

Reply:
memo is extracted to take root and execute pallet ibc extrinsic and pass the memo that inside 'next' field of current extracted memo. so we still need to serialise to pass new(unwraped memo) further.

Reply:
Why we cannot just pass the given `.next` memo? Without any (de)serialization

Reply:
If it's a different type, just add `T::MemoMessage: From<MemoData>` (or `TryFrom`) bound. I don't see any reason to use `serde_json` in this case

Reply:
```
mismatched types
expected enum `std::option::Option<<T as pallet::Config>::MemoMessage>`
   found enum `std::option::Option<&MemoData>`
```
I need to cast Memodata to T::MemoMessage type and it should be some json representation to be compatible on every chain to execute successfully  every hop.

Reply:
Add `T::MemoMessage: From<MemoData>` and use `x.into()`

Reply:
composable repo has his own MemoData struct for some XCVM integration.
so composable uses their own struct that equal to centauri MemoData..

Reply:
Ok, what's the problem?

Reply:
Btw, if we use the same structure twice, it sounds like we can reuse one from the other side. 
And also, do we really need the `MemoMessage` type to be generic? Can't we just use one structure for every packet-forward?

Reply:
This is because the same MemoData is using in XCVM and i think XCVM does not want to have a dependency from centauri and centauri are not able to depend on composable because of circuit dependency. 
So i decide to have two similar struct in composable and centauri once they are using only for deserialisation and serialisation.

Reply:
And i would like to keep MessageType Generic,
because this is enum of memo forward, now it is XCM for substrate xcm hop and IBC for cosmos.
And who knows what other enum will be introduced by xcvm, etc.
So we basically can do it non generic but it will require the changes in both repos and several pallets each time when we will extend enum with a different variance.

Reply:
But isn't the MemoData initially comes from IBC? I mean [this](https://github.com/ComposableFi/composable/blob/32fa07505e3002796ee5a9564204c65409be5417/code/parachain/frame/composable-traits/src/xcm/memo.rs#L92) one. In this case, Composable should use the one from pallet-ibc. 

Anyway... Just using `From` trait should be enough for conversion between different types

Reply:
In our case MemoData is our custom type that has a few different options for XCM forward hop and IBC forward hop.

Reply:
Ok, but there shouldn't be any problems with that, no?

Reply:
We tested. It works fine with a different option in memo (XCM, IBC hop) from cosmos chains though Picasso to composable.
Cosmos chain deserialise memo and did not fail and pass memo forward for futures hop.